middle_end/flambda2.0/to_cmm/un_cps.ml

```diff
@@ -1087,27 +1099,104 @@ and apply_cont_trap_actions env e =
       [Cmm.Push cont]
 
 and switch env s =
-  let e, env, _ = simple env (Switch.scrutinee s) in
+  let env, scrutinee, arms = prepare_switch env s in
+  make_switch env scrutinee arms
+
+and prepare_switch env s =
+  (* Regular preparation of switches, simply translate the scrutinee
+     and get the switch arms *)
+  let scrutinee = Switch.scrutinee s in
+  let e, env, _ = simple env scrutinee in
+  let arms = Switch.arms s in
+  (* For binary switches, which can be translated to if-then-elses,
+     it can be interesting to *not* untag the scrutinee (particularly
+     for those coming from source if-then-elses on booleans) as that way
+     the translation can use 2 instructions instead of 3.
+     However, this is only useful to do if the tagged expression is
+     smaller then the untagged one (which is not always true due to
+     arithmetic simplifications performed by cmm_helpers).
+     Additionally for switches with more than 2 arms, not untagging
+     and lifting the switch to perform on tagged integer might be worse
+     (because the discrimant of the arms may not be successive anymore,
+     thus preventing the use of a table), or simply not worth it
+     given the already high number of isntructions needed for big
+     switches (but that might be up-to-debate on small switches with
+     3-5 arms). *)
+  match Target_imm.Map.cardinal arms with
+  | 2 ->
+    begin match match_var_with_extra_info env scrutinee with
+    | None -> env, e, prepare_arms ~lift:false arms
+    | Some Untag e' ->
+      let size_e = cmm_arith_size e in
+      let size_e' = cmm_arith_size e' in
+      if size_e' < size_e then
+        env, e', prepare_arms ~lift:true arms
+      else
+        env, e, prepare_arms ~lift:false arms
+    end
+  | _ -> env, e, prepare_arms ~lift:false arms
+
+and prepare_arms ~lift arms =
+  Target_imm.Map.fold (fun d action acc ->
+    let i = Target_imm.to_targetint' d in
+    let n = if lift then tag_targetint i else i in
+    Targetint.Map.add n action acc
+  ) arms Targetint.Map.empty
+
+and match_var_with_extra_info env simple : Env.extra_info option =
+  Simple.pattern_match simple
+    ~const:(fun _ -> None)
+    ~name:(fun n ->
+      Name.pattern_match n
+        ~symbol:(fun _ -> None)
+        ~var:(fun v ->
+          begin match Env.extra_info env v with
+          | None -> None
+          | Some _ as res -> res
+          end)
+    )
+
+(* Small function to estimate the number of arithmetic instructions
+   in a cmm expression. This is currently used to determine whether
+   untagging an expression resulted in a smaller expresison or not
+   (as can happen because of some arithmetic simplifications performed
+   by cmm_helpers.ml) *)
+and cmm_arith_size e =
+  match (e : Cmm.expression) with
+  | Cop (
+    ( Caddi | Csubi | Cmuli | Cmulhi | Cdivi | Cmodi
+    | Cand | Cor | Cxor | Clsl | Clsr | Casr ), l, _) ->
+    List.fold_left (+) 1 @@
+    List.map cmm_arith_size l
+  | _ -> 0
```

Comment:
Admittedly, this function can not be used for generic size measurement, but rather specifically to check whether some simplification of the expression has been done at the root of the expression. Ideally, rather than this single-argument function for size, the code should use a comparison function that compare the sizes of two expression by stopping as soon as common sub-expressions are encountered, but that would be too complicated I think.

Actually, in that context, we know that there will be common sub-expressions between the expression whose sizes we will compare, and even more any expression whose head is not an arithmetic operation will necessarily be common to both expression whose size is compared, thus not counting those in the size estimation is correct wrt to the size comparison and saves time (though it might be a good thing to document that in the comments). This is why deeply nested arithmetic expression (such as the ones below a load, or an other operation) are not counted.

Reply:
The default case cannot raise (because it will occur naturally, e.g. for ints just loaded), but can be given any value: it does not matter much since any non-arithmetic expressions should be common to both expressions being compared.